### DISCUSSION TITLE
Google Tag Manager: There is a way to integrate GTM with material-mkdocs

### DISCUSSION THREAD

**Discussion by Marlony-Gabriel** [squidfunk/mkdocs-material#5607] [Q&A]
Hello.
I'm trying to finda a way to integrate the GTM from GA4 in my documentation project.
My problem is that I build the project with gitlab runners directly on the server. And for what I saw, you need to add some code in your html pages to getting it work.
There is a way to do this, without building the project first and adding the code lines, before sent it to my server ?

**Comment by squidfunk** [comment]
Yes, several users already did that successfully. Please try and use the issue/discussion search, as this was asked before.

**Reply by Marlony-Gabriel** [reply]
Hi @squidfunk , thanks for your time.
I'm not so good in the matter of codding yet. I search issues and discussion, but I still don't understand one thing:
1 - If my mkdocs.yml already has this lines: 
    
```
extra:
  analytics:
    provider: google
    property: G-XXXXXXXXXX
```

It's already integrate with tag manager (GTM-XXXXXXX) or I still need to override base.html or main.html (or any other page) and includes the code lines below
in the "head" section: 

```
<!-- Google Tag Manager -->
<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','GTM-XXXXXXX');</script>
<!-- End Google Tag Manager -->
```

And in the "body" section
```
<!-- Google Tag Manager (noscript) -->
<noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
<!-- End Google Tag Manager (noscript) -->
```
Again, thank to take a time and answer me. Mkdocs-material is awesome and I love it.

**Reply by squidfunk** [reply]
No, we don't integrate GTM natively, only Google Analytics 4. With the sunset of Universal Analytics, we might reevaluate in the future, but it's currently not a priority. Thus, you would need to provide a custom analytics integration, [as described in our docs](https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-analytics/#custom-site-analytics).

**Reply by rachel-rigdon-sp** [reply]
I'd like to +1 for incorporating GTM natively. It's especially helpful for large enterprises.

**Reply by squidfunk** [reply]
Thanks for the feedback! Since we've thrown out UA, we're open to PRs to integrate GTM. It shouldn't be too hard to integrate this functionality, but it's currently not a priority, given how many other features we have in our pipeline ☺️

**Reply by ooker777** [reply]
have you successfully integrated GTM? Can you post your final code?

**Reply by rachel-rigdon-sp** [reply]
@ooker777 Unfortunately, no. I'm just a tech writer, so I'm having to work with my web team to get the script written. Once we get that sorted, I'll update this post. 

If anyone else has resolved it and can share their code, that would be a huge help.

**Reply by normandy7** [reply]
I'm also a tech writer and by no means frontend-savvy, but we use multiple providers and I got GTM to work like this:

1. Add the following to `overrides/partials/integrations/analytics/custom.html`:

  ```html
  <!-- Google Tag Manager -->
  <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','GTM-XXXXXXX');</script>
  <!-- End Google Tag Manager -->
  ```

2. Add the following to `overrides/main.html`:

  ```html
  {% extends "base.html" %}
  
  {% block scripts %}
    {{ super() }}
    
    <!-- Google Tag Manager (noscript) -->
    <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
    height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <!-- End Google Tag Manager (noscript) -->
  {% endblock %}
  ```

**Reply by kevinah95** [reply]
> I'm also a tech writer and by no means frontend-savvy, but we use multiple providers and I got GTM to work like this:
> 
> 1. Add the following to `overrides/partials/integrations/analytics/custom.html`:
> 
> ```
> 
> <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
> new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
> j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
> 'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
> })(window,document,'script','dataLayer','GTM-XXXXXXX');</script>
> 
> ```
> 
> 2. Add the following to `overrides/main.html`:
> 
> ```
> {% extends "base.html" %}
> 
> {% block scripts %}
>   {{ super() }}
>   
>   
>   <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
>   height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
>   
> {% endblock %}
> ```



> I'm also a tech writer and by no means frontend-savvy, but we use multiple providers and I got GTM to work like this:
> 
> 1. Add the following to `overrides/partials/integrations/analytics/custom.html`:
> 
> ```
> 
> <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
> new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
> j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
> 'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
> })(window,document,'script','dataLayer','GTM-XXXXXXX');</script>
> 
> ```
> 
> 2. Add the following to `overrides/main.html`:
> 
> ```
> {% extends "base.html" %}
> 
> {% block scripts %}
>   {{ super() }}
>   
>   
>   <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
>   height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
>   
> {% endblock %}
> ```

Don't forget to modify mkdocs.yml file:

```
extra:
  analytics:
    provider: custom
```

**Comment by tomfast** [comment]
This was :1st_place_medal: . Thanks all this @normandy7 and @kevinah95 !

**Comment by oleg-nenashev** [comment]
@squidfunk  would you be open to a PR that adds the stock implementation to Material for MkDocs? It's kind of a good first issue, but GTM gets trendy

**Reply by squidfunk** [reply]
Currently no, sorry. The reason is we're working on [fundamental changes](https://squidfunk.github.io/mkdocs-material/blog/2024/08/19/how-were-transforming-material-for-mkdocs/) right now. In the end, integrations like this should be much simpler and should be achievable in a much more modular way. Additionally, this is a prime example of customizations, maybe even via plugins or theme extension.

**Comment by vikian050194** [comment]
Hello! Probably there is someone else who interested in temp solution till fundamental changes will be released.
There is a bit shorter solution:
1. Just update `mkdocs.yml`
```
...
theme:
  name: material
  custom_dir: docs/overrides
...
```
2. Add `docs/overrides/main.html`
```html
{% extends "base.html" %}

{% block analytics %}
    <!-- Google Tag Manager -->
    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
    new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
    j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
    'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
    })(window,document,'script','dataLayer','GTM-XXXXXXXX');</script>
    <!-- End Google Tag Manager -->

    {{ super() }}
{% endblock %}

{% block header %}
    <!-- Google Tag Manager (noscript) -->
    <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXXX"
    height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <!-- End Google Tag Manager (noscript) -->

    {{ super() }}
{% endblock %}
```

Thank you, @teners for this solution example
